### ISSUE TITLE
arrayHelpers remove deletes last Item

### ISSUE THREAD

**Issue by robinthyssen** [jaredpalmer/formik#1504]
### Current Behavior
When using arrayHelpers in a deeply nested Formik structure arrayHelpers.remove ignores the provided index and deletes the last item in the FieldArray instead.

Button that calls arrayHelpers.remove that is passed in via props
![image](https://user-images.githubusercontent.com/32195729/57538126-062d9180-7348-11e9-959e-4c9b45ef4871.png)
Screenshot from devtools showing that the correct index is passed
![image](https://user-images.githubusercontent.com/32195729/57538138-0cbc0900-7348-11e9-89bc-2f1c1b9c2f00.png)

### Expected behavior
arrayHelpers functioning as usual. 

### Reproducible example
I don't have a reproducable example as my problem only arises when the FieldArray is deeply nested.

### Additional context
In my usecase I receive a JSON Object that stores Display Settings for various devices. 
The part shown below is an Array of Key-Value pairs for different display settings (eg. 
{spacing_s: 8px}. 

FieldArray that works:
```
class SelectionContainer extends React.Component {
    constructor(props) {
        super(props)

    }

    render() {
        var device = this.props.device;
        var index = this.props.index;
        var prefix = this.props.prefix;
        return (
            <FieldArray
                name={`block_form_settings.${prefix}.${device}.settings.${index}.selection`}
                render={arrayHelpers => {
                    if (!_.isObject(arrayHelpers.form.values.block_form_settings[prefix][device].settings[index].selection)) {
                        arrayHelpers.push({})
                        return (<div></div>)
                    }
                    else {
                        return (
                            <div className={'june_settings-form__option-list'}>
                                {arrayHelpers.form.values.block_form_settings[prefix][device].settings[index].selection.map((selection, mapIndex) => {
                                    var isLast = () => {
                                        return mapIndex + 1 == arrayHelpers.form.values.block_form_settings[prefix][device].settings[this.props.index].selection.length
                                    }
                                    return (<KeyValuePair
                                        setting={this.props.setting}
                                        index={this.props.index}
                                        device={this.props.device}
                                        childKey={mapIndex}
                                        moveUp={arrayHelpers.swap}
                                        moveDown={arrayHelpers.swap}
                                        removeKey={arrayHelpers.remove}
                                        currentPair={selection}
                                        form={arrayHelpers.form}
                                        isLast={isLast()}
                                        prefix={prefix}
                                        formikProps={this.props.formikProps}
                                        key={`${this.props.setting.type}.${this.props.index}.${mapIndex}`}

                                    />)
                                })}
                                <div className={'june_settings-form__option-add'}>
                                    <Button
                                        onClick={() =>{ arrayHelpers.push({})}}
                                        label={'Add New Key'}
                                        buttonStyle={'june_button--ghost june_button--small'}
                                    />
                                </div>
                            </div>
                        )
                    }
                }
                }
            />
        )
    }
}
```

Similar FieldArray that is just more deeply nested. The only difference here is the name/path that is provided to Formik.
```
class SelectionContainerSuffix extends React.Component {
    constructor(props) {
        super(props)

    }
   render() {
        var device = this.props.device;
        var index = this.props.index;
        var suffix;
        var direction;

        if (this.props.suffix != undefined) {
            suffix = `${this.props.suffix}`
        } else {
            suffix = '';
        }
        if (this.props.direction != undefined) {
            direction = `${this.props.direction}`
        } else {
            direction = '';
        }
        var prefix = this.props.prefix;
        return (
            <FieldArray
                name={`block_form_settings.${prefix}.${device}.settings.${index}.selection.${suffix}.${direction}`}
                render={arrayHelpers => {
                    if (_.get(arrayHelpers.form.values.block_form_settings[prefix][device].settings[index].selection, suffix, -1) == -1) {
                        arrayHelpers.push({})
                        return (<div></div>)
                    }
                    else {
                        return (
                            <div className={'june_settings-form__option-list'}>
                               {arrayHelpers.form.values.block_form_settings[prefix][device].settings[index].selection[suffix][direction].map((selection, mapIndex) => {
                                    var isLast = () => {
                                        return mapIndex + 1 == arrayHelpers.form.values.block_form_settings[prefix][device].settings[this.props.index].selection[suffix][direction].length
                                    }
                                    return (<KeyValuePair
                                        setting={this.props.setting}
                                        index={this.props.index}
                                        device={this.props.device}
                                        childKey={mapIndex}
                                        moveUp={arrayHelpers.swap}
                                        moveDown={arrayHelpers.swap}
                                        removeKey={arrayHelpers.remove}
                                        currentPair={selection}
                                        form={arrayHelpers.form}
                                        isLast={isLast()}
                                        prefix={prefix}
                                        formikProps={this.props.formikProps}
                                        key={`${this.props.setting.type}.${this.props.index}.${mapIndex}`}
                                        suffix={this.props.suffix}
                                        direction={this.props.direction}
                                    />)
                                })}
                                <div className={'june_settings-form__option-add'}>
                                    <Button
                                        onClick={() =>{ arrayHelpers.push({})}}
                                        label={'Add New Key'}
                                        buttonStyle={'june_button--ghost june_button--small'}
                                    />
                                </div>
                            </div>
                        )
                    }
                }
                }
            />
        )
    }
}
```

### Your environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Formik           | 1.5.4
| React            | 16.6.3
| TypeScript       | 3.3.3333
| Browser          | 73.0.3683.103
| npm         | 6.4.1
| Operating System | Ubuntu Mate 18.04 / MacOS

**Comment by jukkhop** [comment]
Having this problem as well.

**Comment by robinthyssen** [comment]
Issue resolved! I messed myself up in shouldComponentUpdate in my custom Formik Component.

**Comment by robinthyssen** [comment]
Issue resolved! I messed myself up in shouldComponentUpdate in my custom Formik Component.

**Comment by vincentbello** [comment]
Hi @robinthyssen, can you clarify exactly what you did in `shouldComponentUpdate` that broke the remove functionality? I'm encountering exactly the same issue.

**Comment by jackkranz** [comment]
Did anyone determine the solution to this?
@robinthyssen  @vincentbello

**Comment by jukkhop** [comment]
Yeah, I had the exact same issue and was able to resolve it.

I was not properly passing the value from the Formik render props to the <input> component. So the "data binding" was working only one way, from DOM to Formik. And when I tried to manipulate the state with arrayHelpers (from FieldArray render props), it resulted to Formik's state and the DOM becoming unsynchronized with each other. 

And to explain the same thing in code:

```
<Formik
  // ...
  render={({
    handleBlur,
    handleChange,
    handleSubmit,
    values,
  }) => (
    <Form onSubmit={handleSubmit}>
      <FieldArray
        name="users"
        render={array => (
          <Fragment>
            {values.users.map((user, index) => (
              <Input
                id={`users.${index}.email`}
                name={`users.${index}.email`}
                onBlur={handleBlur}
                onChange={handleChange}
                type="text"
                value={user.email}
              />
            ))}
          </Fragment>
        )}
      />
    </Form>
  )}
/>
```

Basically I just added the `value={user.email}`

**Comment by kkerley** [comment]
@jackkranz can you provide any additional information about how you fixed this? I've run into the same problem and am using the latest Formik 2.1.4 along with the latest versions of React and React DOM. There was no change in behavior from 1.5.8 to 2.1.4 for me and I can't figure out why. My wrapper div around the repeating field group has its key set to the array index. 

What's especially weird is that of the three fields in my repeating group (song title, set order, and notes), when I remove, say, the third song in the set list, the song title doesn't change but the set order will go from 3 to 4 and all of the notes shift up one row but the song titles all remain the same with the very last song title disappearing.

This is so weird and I can't figure out why it's happening. I'm using an Autosuggest component based on react-autosuggest for song title and that's the only thing I can think of that would be causing an issue but I can't see how.

**Comment by jackkranz** [comment]
@kkerley . Sounds like your issue is related to using the array index as the key. When you delete an item from the array React  will not rerender the div because the array index (your key) hasn't changed the data at that index has changed.

You need to use a unique value for the key that way React knows when to re-render the div.  if your song doesn't have a unique identifier you can try using `JSON.stringify(YourSongObject)`.

**Comment by kkerley** [comment]
@jackkranz THANK YOU. I just implemented that and it's working perfectly.

**Comment by kkerley** [comment]
FYI for anyone coming to this and applying the above fix...while using JSON.stringify(object) fixes the issue with arrayHelpers.remove, it also introduced a nasty bug that would cause any field inside of a fieldArray to lose focus after typing one character ([more on this here](https://github.com/jaredpalmer/formik/issues/505)).

To fix this, I changed my key to be ${index}-${song.song_slug}-${song.song_id}

The arrayHelpers.remove() function still works as it should and the issue with typing in fields inside of the fieldArray was fixed as well.

**Comment by saisiddulugari** [comment]
I faced a similar issue with `arrayHelpers.remove(index)`, using`arrayHelpers.handleRemove(index)`and it seems to work.
